File: backend/services/UserServices.js
/**
 * @author Bodo Desderio <rooiboktechltd@gmail.com>
 * @copyright 2026 Rooibok Technologies. All rights reserved.
 */
import { pool } from '../db/index.js';
import { sendPushNotificationAsync } from './NotificationService.js';
import { logger } from '../utils/logger.js';

const ALLOWED_USER_SORT_FIELDS = ['id', 'created_at', 'first_name', 'last_name', 'email'];
const ALLOWED_CONSULTANT_SORT_FIELDS = ['id', 'created_at', 'rating', 'first_name', 'last_name'];
const ALLOWED_SORT_ORDERS = ['ASC', 'DESC'];

// ─── Profile ──────────────────────────────────────────────────────────────────
export const getProfile = async (userId, role) => {
    try {
        let query;
        if (role === 'consultant') {
            query = `SELECT u.id, u.email, u.role, u.is_active, u.created_at,
                            p.username, p.first_name, p.last_name, p.profile_image, p.phone, p.dob, p.gender, p.push_token, p.notifications_enabled,
                            cd.profession, cd.experience, cd.rating, cd.education, cd.language, cd.available_from, cd.available_to, cd.available_days, cd.is_approved
                     FROM users u
                     LEFT JOIN profiles p ON p.user_id = u.id
                     LEFT JOIN consultant_details cd ON cd.user_id = u.id
                     WHERE u.id = $1 AND u.deleted_at IS NULL`;
        } else {
            query = `SELECT u.id, u.email, u.role, u.is_active, u.created_at,
                            p.username, p.first_name, p.last_name, p.profile_image, p.phone, p.dob, p.gender, p.push_token, p.notifications_enabled
                     FROM users u
                     LEFT JOIN profiles p ON p.user_id = u.id
                     WHERE u.id = $1 AND u.deleted_at IS NULL`;
        }

        const { rows } = await pool.query(query, [userId]);

        if (!rows || rows.length === 0) {
            return { success: false, message: 'User profile not found.' };
        }

        return { success: true, message: 'User profile retrieved.', user: rows[0] };
    } catch (error) {
        logger.error({ error }, 'getProfile failed');
        return { success: false, message: 'Profile not found. Please try again later.' };
    }
};

// ─── User Details ─────────────────────────────────────────────────────────────
export const getUserDetails = async (userId) => {
    try {
        const { rows } = await pool.query(
            `SELECT u.id, u.email, u.role, u.created_at,
                    p.username, p.first_name, p.last_name
             FROM users u
             LEFT JOIN profiles p ON p.user_id = u.id
             WHERE u.id = $1 AND u.deleted_at IS NULL`,
            [userId]
        );
        if (rows.length === 0) return { success: false, message: 'User not found' };
        return { success: true, message: 'User details fetched.', user: rows[0] };
    } catch (error) {
        logger.error({ error }, 'getUserDetails failed');
        return { success: false, message: 'Failed to fetch user data.' };
    }
};

// ─── Consultant Details ───────────────────────────────────────────────────────
export const getConsultantDetails = async (consultantId) => {
    try {
        const { rows } = await pool.query(
            `SELECT u.id, u.email, u.role, u.created_at,
                    p.username, p.first_name, p.last_name, p.profile_image,
                    cd.profession, cd.experience, cd.rating, cd.education, cd.language,
                    cd.available_from, cd.available_to, cd.available_days
             FROM users u
             LEFT JOIN profiles p ON p.user_id = u.id
             LEFT JOIN consultant_details cd ON cd.user_id = u.id
             WHERE u.id = $1 AND u.role = 'consultant' AND u.deleted_at IS NULL`,
            [consultantId]
        );
        if (rows.length === 0) return { success: false, message: 'Consultant not found' };
        return { success: true, message: 'Consultant details fetched.', consultant: rows[0] };
    } catch (error) {
        logger.error({ error }, 'getConsultantDetails failed');
        return { success: false, message: 'Failed to fetch consultant data.' };
    }
};

// ─── List Consultants ─────────────────────────────────────────────────────────
export const getConsultants = async ({ page = 1, limit = 10, profession, search, sortBy = 'id', sortOrder = 'DESC' } = {}) => {
    try {
        const safeSortBy = ALLOWED_CONSULTANT_SORT_FIELDS.includes(sortBy) ? sortBy : 'id';
        const safeSortOrder = ALLOWED_SORT_ORDERS.includes(sortOrder.toUpperCase()) ? sortOrder.toUpperCase() : 'DESC';
        const offset = (page - 1) * limit;

        // Map sort fields to correct table aliases
        const sortFieldMap = { id: 'u.id', created_at: 'u.created_at', rating: 'cd.rating', first_name: 'p.first_name', last_name: 'p.last_name' };
        const sortColumn = sortFieldMap[safeSortBy] || 'u.id';

        let paramIdx = 1;
        const queryParams = [];
        const whereClauses = [`u.role = 'consultant'`, `u.deleted_at IS NULL`];

        if (profession) {
            whereClauses.push(`cd.profession = $${paramIdx++}`);
            queryParams.push(profession);
        }
        if (search) {
            whereClauses.push(`(p.first_name ILIKE $${paramIdx} OR p.last_name ILIKE $${paramIdx + 1} OR u.email ILIKE $${paramIdx + 2})`);
            const s = `%${search}%`;
            queryParams.push(s, s, s);
            paramIdx += 3;
        }

        const whereClause = 'WHERE ' + whereClauses.join(' AND ');

        const query = `SELECT u.id, u.email, u.role, u.created_at,
                              p.username, p.first_name, p.last_name, p.profile_image,
                              cd.profession, cd.experience, cd.rating, cd.education, cd.language,
                              cd.available_from, cd.available_to, cd.available_days
                       FROM users u
                       LEFT JOIN profiles p ON p.user_id = u.id
                       LEFT JOIN consultant_details cd ON cd.user_id = u.id
                       ${whereClause}
                       ORDER BY ${sortColumn} ${safeSortOrder}
                       LIMIT $${paramIdx++} OFFSET $${paramIdx++}`;
        queryParams.push(Number(limit), Number(offset));

        const { rows: consultants } = await pool.query(query, queryParams);

        const countParams = queryParams.slice(0, -2);
        const { rows: countResult } = await pool.query(
            `SELECT COUNT(*) as total FROM users u
             LEFT JOIN profiles p ON p.user_id = u.id
             LEFT JOIN consultant_details cd ON cd.user_id = u.id
             ${whereClause}`,
            countParams
        );

        const total = parseInt(countResult[0].total);
        return {
            success: true,
            message: 'Fetched consultants',
            consultants,
            pagination: { page: Number(page), limit: Number(limit), total, totalPages: Math.ceil(total / limit) },
        };
    } catch (error) {
        logger.error({ error }, 'getConsultants failed');
        return { success: false, message: 'Failed to fetch consultants data.' };
    }
};

// ─── List Users ───────────────────────────────────────────────────────────────
export const getUsers = async ({ page = 1, limit = 10, role, search, sortBy = 'id', sortOrder = 'DESC' } = {}) => {
    try {
        const safeSortBy = ALLOWED_USER_SORT_FIELDS.includes(sortBy) ? sortBy : 'id';
        const safeSortOrder = ALLOWED_SORT_ORDERS.includes(sortOrder.toUpperCase()) ? sortOrder.toUpperCase() : 'DESC';
        const offset = (page - 1) * limit;

        const sortFieldMap = { id: 'u.id', created_at: 'u.created_at', first_name: 'p.first_name', last_name: 'p.last_name', email: 'u.email' };
        const sortColumn = sortFieldMap[safeSortBy] || 'u.id';

        let paramIdx = 1;
        const queryParams = [];
        const whereClauses = ['u.deleted_at IS NULL'];

        if (role) {
            whereClauses.push(`u.role = $${paramIdx++}`);
            queryParams.push(role);
        }
        if (search) {
            whereClauses.push(`(p.first_name ILIKE $${paramIdx} OR p.last_name ILIKE $${paramIdx + 1} OR u.email ILIKE $${paramIdx + 2})`);
            const s = `%${search}%`;
            queryParams.push(s, s, s);
            paramIdx += 3;
        }

        const whereClause = 'WHERE ' + whereClauses.join(' AND ');

        const query = `SELECT u.id, u.email, u.role, u.created_at, u.updated_at,
                              p.username, p.first_name, p.last_name, p.profile_image, p.dob, p.gender
                       FROM users u
                       LEFT JOIN profiles p ON p.user_id = u.id
                       ${whereClause}
                       ORDER BY ${sortColumn} ${safeSortOrder}
                       LIMIT $${paramIdx++} OFFSET $${paramIdx++}`;
        queryParams.push(Number(limit), Number(offset));

        const { rows: users } = await pool.query(query, queryParams);

        const countParams = queryParams.slice(0, -2);
        const { rows: countResult } = await pool.query(
            `SELECT COUNT(*) as total FROM users u
             LEFT JOIN profiles p ON p.user_id = u.id
             ${whereClause}`,
            countParams
        );

        const total = parseInt(countResult[0].total);
        return {
            success: true,
            message: 'Fetched users',
            users,
            pagination: { page: Number(page), limit: Number(limit), total, totalPages: Math.ceil(total / limit) },
        };
    } catch (error) {
        logger.error({ error }, 'getUsers failed');
        return { success: false, message: 'Failed to fetch users data.' };
    }
};

// ─── Delete ───────────────────────────────────────────────────────────────────
export const deleteUser = async (userId) => {
    try {
        // Soft delete
        const result = await pool.query('UPDATE users SET deleted_at = NOW() WHERE id = $1 AND deleted_at IS NULL', [userId]);
        if (result.rowCount === 0) return { success: false, message: 'User not found' };
        return { success: true, message: 'User deleted.' };
    } catch (error) {
        logger.error({ error }, 'deleteUser failed');
        return { success: false, message: 'Failed to delete user.' };
    }
};

export const deleteConsultant = async (consultantId) => {
    try {
        // Soft delete — consultants are in the users table
        const result = await pool.query(
            `UPDATE users SET deleted_at = NOW() WHERE id = $1 AND role = 'consultant' AND deleted_at IS NULL`,
            [consultantId]
        );
        if (result.rowCount === 0) return { success: false, message: 'Consultant not found' };
        return { success: true, message: 'Consultant deleted.' };
    } catch (error) {
        logger.error({ error }, 'deleteConsultant failed');
        return { success: false, message: 'Failed to delete consultant.' };
    }
};

// ─── Update Profile ───────────────────────────────────────────────────────────
export const updateProfile = async (userId, role, fieldsToUpdate) => {
    try {
        const PROFILE_FIELDS = ['username', 'first_name', 'last_name', 'phone', 'profile_image', 'dob', 'gender'];
        const USER_FIELDS = ['email'];
        const profileUpdates = [];
        const profileValues = [];
        const userUpdates = [];
        const userValues = [];
        let profileParamIdx = 1;
        let userParamIdx = 1;

        Object.entries(fieldsToUpdate).forEach(([key, value]) => {
            if (value !== undefined && value !== null && value !== '') {
                if (PROFILE_FIELDS.includes(key)) {
                    profileUpdates.push(`${key} = $${profileParamIdx++}`);
                    profileValues.push(value);
                } else if (USER_FIELDS.includes(key)) {
                    userUpdates.push(`${key} = $${userParamIdx++}`);
                    userValues.push(value);
                }
            }
        });

        if (profileUpdates.length === 0 && userUpdates.length === 0) {
            return { success: false, message: 'No valid fields to update.' };
        }

        let updated = false;

        if (profileUpdates.length > 0) {
            profileValues.push(userId);
            const result = await pool.query(
                `UPDATE profiles SET ${profileUpdates.join(', ')} WHERE user_id = $${profileParamIdx}`,
                profileValues
            );
            if (result.rowCount > 0) updated = true;
        }

        if (userUpdates.length > 0) {
            userUpdates.push(`updated_at = NOW()`);
            userValues.push(userId);
            const result = await pool.query(
                `UPDATE users SET ${userUpdates.join(', ')} WHERE id = $${userParamIdx + 1}`,
                userValues
            );
            if (result.rowCount > 0) updated = true;
        }

        if (!updated) {
            return { success: false, message: 'No changes made.' };
        }

        const updatedProfile = await getProfile(userId, role);
        return { success: true, message: 'Profile updated successfully.', user: updatedProfile.user };
    } catch (error) {
        logger.error({ error }, 'updateProfile failed');
        return { success: false, message: 'Failed to update profile.' };
    }
};

// ─── Push Tokens ─────────────────────────────────────────────────────────────
export const saveUserPushToken = async (userId, pushToken) => {
    await pool.query('UPDATE profiles SET push_token = $1 WHERE user_id = $2', [pushToken, userId]);
};

export const saveConsultantPushToken = async (consultantId, pushToken) => {
    await pool.query('UPDATE profiles SET push_token = $1 WHERE user_id = $2', [pushToken, consultantId]);
};

// ─── Push Notifications ───────────────────────────────────────────────────────
export const sendPushNotificationToUser = async (userId, title, body, data) => {
    try {
        const { rows } = await pool.query(
            'SELECT push_token, notifications_enabled FROM profiles WHERE user_id = $1',
            [userId]
        );
        if (!rows.length || !rows[0].push_token || rows[0].notifications_enabled === false) {
            return { success: false, message: 'Push notifications disabled or token not found.' };
        }
        const result = await sendPushNotificationAsync(rows[0].push_token, title, body, data);
        await saveNotification({ recipientId: userId, title, body, data });
        return result;
    } catch (error) {
        logger.error({ error }, 'sendPushNotificationToUser failed');
        return { success: false, message: 'Failed to send notification.' };
    }
};

export const sendPushNotificationToConsultant = async (consultantId, title, body, data) => {
    try {
        const { rows } = await pool.query(
            'SELECT push_token, notifications_enabled FROM profiles WHERE user_id = $1',
            [consultantId]
        );
        if (!rows.length || !rows[0].push_token || rows[0].notifications_enabled === false) {
            return { success: false, message: 'Push notifications disabled or token not found.' };
        }
        const result = await sendPushNotificationAsync(rows[0].push_token, title, body, data);
        await saveNotification({ recipientId: consultantId, title, body, data });
        return result;
    } catch (error) {
        logger.error({ error }, 'sendPushNotificationToConsultant failed');
        return { success: false, message: 'Failed to send notification.' };
    }
};

// ─── Notifications ────────────────────────────────────────────────────────────
export const saveNotification = async ({ recipientId, title, body, data }) => {
    await pool.query(
        'INSERT INTO notifications (recipient_id, title, body, data) VALUES ($1, $2, $3, $4)',
        [recipientId, title, body, JSON.stringify(data || {})]
    );
};

export const getNotificationsForUser = async (userId) => {
    const { rows } = await pool.query(
        'SELECT * FROM notifications WHERE recipient_id = $1 ORDER BY created_at DESC LIMIT 50',
        [userId]
    );
    return rows;
};

export const markNotificationAsRead = async (notificationId, userId) => {
    await pool.query(
        'UPDATE notifications SET is_read = TRUE WHERE id = $1 AND recipient_id = $2',
        [notificationId, userId]
    );
};

export const updateUserNotificationPreference = async (userId, enabled) => {
    await pool.query('UPDATE profiles SET notifications_enabled = $1 WHERE user_id = $2', [enabled, userId]);
    return { success: true, message: 'Notification preference updated.' };
};

export const updateConsultantNotificationPreference = async (consultantId, enabled) => {
    await pool.query('UPDATE profiles SET notifications_enabled = $1 WHERE user_id = $2', [enabled, consultantId]);
    return { success: true, message: 'Notification preference updated.' };
};

File: services/django-api/accounts/user_services.py
# @author Bodo Desderio <rooiboktechltd@gmail.com>
# @copyright 2026 Rooibok Technologies. All rights reserved.
"""
User-domain payload builders and list queries, ported from UserServices.js.

Response keys mirror the Node SELECT columns exactly (cutover parity). Push delivery
lives in R6 (Celery); the notification-send endpoints here only persist rows.
"""
from __future__ import annotations

import math

from django.db.models import Q

from content.models import Notification

from .models import ConsultantDetails, Profile, Role, User

CONSULTANT_SORT = {
    "id": "id",
    "created_at": "created_at",
    "rating": "consultant_detail__rating",
    "first_name": "profile__first_name",
    "last_name": "profile__last_name",
}
USER_SORT = {
    "id": "id",
    "created_at": "created_at",
    "first_name": "profile__first_name",
    "last_name": "profile__last_name",
    "email": "email",
}


def _iso(v):
    return v.isoformat() if v is not None else None


def _profile(user_id) -> Profile | None:
    return Profile.objects.filter(user_id=user_id).first()


def profile_payload(user: User) -> dict:
    """Full self-profile (getProfile) — user + profile (+ consultant details if applicable)."""
    p = _profile(user.id)
    data = {
        "id": str(user.id),
        "email": user.email,
        "role": user.role,
        "is_active": user.is_active,
        "created_at": _iso(user.created_at),
        "username": getattr(p, "username", None),
        "first_name": getattr(p, "first_name", None),
        "last_name": getattr(p, "last_name", None),
        "profile_image": getattr(p, "profile_image", None),
        "phone": getattr(p, "phone", None),
        "dob": _iso(getattr(p, "dob", None)),
        "gender": getattr(p, "gender", None),
        "push_token": getattr(p, "push_token", None),
        "notifications_enabled": getattr(p, "notifications_enabled", None),
    }
    if user.role == Role.CONSULTANT:
        cd = ConsultantDetails.objects.filter(user=user).first()
        if cd is not None:
            data.update(
                {
                    "profession": cd.profession,
                    "experience": cd.experience,
                    "rating": cd.rating,
                    "education": cd.education,
                    "language": cd.language,
                    "available_from": _iso(cd.available_from),
                    "available_to": _iso(cd.available_to),
                    "available_days": cd.available_days,
                    "is_approved": cd.is_approved,
                }
            )
    return data


def user_detail_payload(user: User) -> dict:
    p = _profile(user.id)
    return {
        "id": str(user.id),
        "email": user.email,
        "role": user.role,
        "created_at": _iso(user.created_at),
        "username": getattr(p, "username", None),
        "first_name": getattr(p, "first_name", None),
        "last_name": getattr(p, "last_name", None),
    }


def consultant_detail_payload(user: User, cd: ConsultantDetails | None = None, p: Profile | None = None) -> dict:
    p = p or _profile(user.id)
    cd = cd or ConsultantDetails.objects.filter(user=user).first()
    return {
        "id": str(user.id),
        "email": user.email,
        "role": user.role,
        "created_at": _iso(user.created_at),
        "username": getattr(p, "username", None),
        "first_name": getattr(p, "first_name", None),
        "last_name": getattr(p, "last_name", None),
        "profile_image": getattr(p, "profile_image", None),
        "profession": getattr(cd, "profession", None),
        "experience": getattr(cd, "experience", None),
        "rating": getattr(cd, "rating", None),
        "education": getattr(cd, "education", None),
        "language": getattr(cd, "language", None),
        "available_from": _iso(getattr(cd, "available_from", None)),
        "available_to": _iso(getattr(cd, "available_to", None)),
        "available_days": getattr(cd, "available_days", None),
    }


def _page_limit(params, default_limit=10):
    try:
        page = max(int(params.get("page", 1)), 1)
    except (TypeError, ValueError):
        page = 1
    try:
        limit = max(int(params.get("limit", default_limit)), 1)
    except (TypeError, ValueError):
        limit = default_limit
    return page, limit


def _order(params, sort_map, default="id"):
    by = params.get("sortBy", default)
    by = by if by in sort_map else default
    order = (params.get("sortOrder") or "DESC").upper()
    prefix = "" if order == "ASC" else "-"
    return prefix + sort_map[by]


def list_consultants(params) -> dict:
    qs = User.objects.filter(role=Role.CONSULTANT, deleted_at__isnull=True)
    if params.get("profession"):
        qs = qs.filter(consultant_detail__profession=params["profession"])
    if params.get("search"):
        s = params["search"]
        qs = qs.filter(Q(profile__first_name__icontains=s) | Q(profile__last_name__icontains=s) | Q(email__icontains=s))
    qs = qs.distinct().order_by(_order(params, CONSULTANT_SORT))
    total = qs.count()
    page, limit = _page_limit(params)
    offset = (page - 1) * limit
    consultants = [consultant_detail_payload(u) for u in qs[offset : offset + limit]]
    return {
        "success": True,
        "message": "Fetched consultants",
        "consultants": consultants,
        "pagination": {"page": page, "limit": limit, "total": total, "totalPages": math.ceil(total / limit)},
    }


def list_users(params) -> dict:
    qs = User.objects.filter(deleted_at__isnull=True)
    if params.get("role"):
        qs = qs.filter(role=params["role"])
    if params.get("search"):
        s = params["search"]
        qs = qs.filter(Q(profile__first_name__icontains=s) | Q(profile__last_name__icontains=s) | Q(email__icontains=s))
    qs = qs.distinct().order_by(_order(params, USER_SORT))
    total = qs.count()
    page, limit = _page_limit(params)
    offset = (page - 1) * limit
    rows = []
    for u in qs[offset : offset + limit]:
        p = _profile(u.id)
        rows.append(
            {
                "id": str(u.id),
                "email": u.email,
                "role": u.role,
                "created_at": _iso(u.created_at),
                "updated_at": _iso(u.updated_at),
                "username": getattr(p, "username", None),
                "first_name": getattr(p, "first_name", None),
                "last_name": getattr(p, "last_name", None),
                "profile_image": getattr(p, "profile_image", None),
                "dob": _iso(getattr(p, "dob", None)),
                "gender": getattr(p, "gender", None),
            }
        )
    return {
        "success": True,
        "message": "Fetched users",
        "users": rows,
        "pagination": {"page": page, "limit": limit, "total": total, "totalPages": math.ceil(total / limit)},
    }


def notification_payload(n: Notification) -> dict:
    return {
        "id": str(n.id),
        "recipient_id": str(n.recipient_id),
        "title": n.title,
        "body": n.body,
        "data": n.data,
        "is_read": n.is_read,
        "created_at": _iso(n.created_at),
    }
